File: go-prepared-app/lib/shared/widgets/community_journey_matcher.dart
import '../../data/models/ai_models.dart';

/// Client-side scoring aligned with API CommunityJourneyMatcher.
int communityInsightScore(CommunityInsightModel insight, JourneyModel journey) {
  final journeyTokens = _journeyTokens(journey);
  final insightTokens = _insightTokens(insight);
  if (journeyTokens.isEmpty || insightTokens.isEmpty) return 0;

  var overlap = 0;
  for (final j in journeyTokens) {
    for (final i in insightTokens) {
      if (_tokensMatch(j, i)) overlap++;
    }
  }
  if (overlap == 0) return 0;

  var score = overlap * 15;
  final location = _token(journey.location);
  if (location.isNotEmpty && insightTokens.contains(location)) score += 35;

  final subtype = _token(journey.journeySubtype);
  final activity = _token(journey.activity);
  if ((subtype.isNotEmpty && insightTokens.contains(subtype)) ||
      (activity.isNotEmpty && insightTokens.contains(activity))) {
    score += 25;
  }

  final type = _token(journey.journeyType);
  if (type.isNotEmpty && insightTokens.contains(type)) score += 15;

  return score > 95 ? 95 : score;
}

List<CommunityInsightModel> filterInsightsForJourney(
  List<CommunityInsightModel> insights,
  JourneyModel journey,
) {
  final scored = insights
      .map((i) => MapEntry(i, communityInsightScore(i, journey)))
      .where((e) => e.value > 0)
      .toList()
    ..sort((a, b) {
      final byScore = b.value.compareTo(a.value);
      if (byScore != 0) return byScore;
      return b.key.votes.compareTo(a.key.votes);
    });
  return scored.map((e) => e.key).toList();
}

String journeyCommunityContext(JourneyModel journey) {
  final parts = <String>[
    if (journey.journeyType != null && journey.journeyType!.isNotEmpty) journey.journeyType!,
    if (journey.journeySubtype != null && journey.journeySubtype!.isNotEmpty) journey.journeySubtype!,
    if (journey.activity != null && journey.activity!.isNotEmpty) journey.activity!,
    if (journey.location != null && journey.location!.isNotEmpty) journey.location!,
  ];
  if (parts.isEmpty) return journey.title;
  return parts.join('|').toUpperCase();
}

Set<String> _journeyTokens(JourneyModel journey) {
  final tokens = <String>{};
  _addTokens(tokens, journey.title);
  _addTokens(tokens, journey.originalQuery);
  _addTokens(tokens, journey.journeyType);
  _addTokens(tokens, journey.journeySubtype);
  _addTokens(tokens, journey.activity);
  _addTokens(tokens, journey.location);
  return tokens.where(_isSignificant).toSet();
}

Set<String> _insightTokens(CommunityInsightModel insight) {
  final tokens = <String>{};
  _addTokens(tokens, insight.journeyContext);
  _addTokens(tokens, insight.title);
  return tokens.where(_isSignificant).toSet();
}

void _addTokens(Set<String> tokens, String? raw) {
  if (raw == null || raw.trim().isEmpty) return;
  for (final part in raw.toLowerCase().split(RegExp(r'[|\s,/\-—]+'))) {
    final t = part.trim();
    if (t.isNotEmpty) tokens.add(t);
  }
}

String _token(String? raw) => raw?.trim().toLowerCase() ?? '';

bool _tokensMatch(String a, String b) {
  if (a == b) return true;
  if (a.length >= 4 && b.length >= 4 && (a.contains(b) || b.contains(a))) return true;
  return false;
}

bool _isSignificant(String token) {
  if (token.length < 2) return false;
  const stop = {'the', 'and', 'for', 'your', 'with', 'from', 'that', 'this', 'day', 'pre', 'prep'};
  return !stop.contains(token);
}

File: go-prepared-api/src/main/java/com/goprepared/api/service/CommunityJourneyMatcher.java
package com.goprepared.api.service;

import com.goprepared.api.domain.CommunityInsightEntity;
import com.goprepared.api.domain.Journey;
import java.util.Arrays;
import java.util.LinkedHashSet;
import java.util.Locale;
import java.util.Set;
import java.util.stream.Collectors;

/** Scores community insights against a journey for hub-scoped feeds. */
final class CommunityJourneyMatcher {

    private CommunityJourneyMatcher() {}

    static int score(CommunityInsightEntity insight, Journey journey) {
        if (insight.getJourney() != null && journey.getId() != null
                && insight.getJourney().getId().equals(journey.getId())) {
            return 100;
        }

        Set<String> journeyTokens = journeyTokens(journey);
        Set<String> insightTokens = insightTokens(insight);
        if (journeyTokens.isEmpty() || insightTokens.isEmpty()) {
            return 0;
        }

        int overlap = overlapCount(journeyTokens, insightTokens);
        if (overlap == 0) {
            return 0;
        }

        int score = overlap * 15;
        if (locationMatch(journey, insightTokens)) {
            score += 35;
        }
        if (subtypeMatch(journey, insightTokens)) {
            score += 25;
        }
        if (typeMatch(journey, insightTokens)) {
            score += 15;
        }
        return Math.min(score, 95);
    }

    private static boolean locationMatch(Journey journey, Set<String> insightTokens) {
        String location = token(journey.getLocation() != null ? journey.getLocation().getName() : null);
        return !location.isEmpty() && insightTokens.contains(location);
    }

    private static boolean subtypeMatch(Journey journey, Set<String> insightTokens) {
        String subtype = token(journey.getJourneySubtype() != null ? journey.getJourneySubtype().getName() : null);
        if (!subtype.isEmpty() && insightTokens.contains(subtype)) {
            return true;
        }
        String activity = token(journey.getActivity() != null ? journey.getActivity().getName() : null);
        return !activity.isEmpty() && insightTokens.contains(activity);
    }

    private static boolean typeMatch(Journey journey, Set<String> insightTokens) {
        String type = token(journey.getJourneyType() != null ? journey.getJourneyType().getName() : null);
        return !type.isEmpty() && insightTokens.contains(type);
    }

    private static int overlapCount(Set<String> journeyTokens, Set<String> insightTokens) {
        int count = 0;
        for (String j : journeyTokens) {
            for (String i : insightTokens) {
                if (tokensMatch(j, i)) {
                    count++;
                }
            }
        }
        return count;
    }

    private static boolean tokensMatch(String a, String b) {
        if (a.equals(b)) {
            return true;
        }
        if (a.length() >= 4 && b.length() >= 4 && (a.contains(b) || b.contains(a))) {
            return true;
        }
        return false;
    }

    private static Set<String> journeyTokens(Journey journey) {
        Set<String> tokens = new LinkedHashSet<>();
        addTokens(tokens, journey.getTitle());
        addTokens(tokens, journey.getOriginalQuery());
        if (journey.getJourneyType() != null) {
            addTokens(tokens, journey.getJourneyType().getName());
        }
        if (journey.getJourneySubtype() != null) {
            addTokens(tokens, journey.getJourneySubtype().getName());
        }
        if (journey.getActivity() != null) {
            addTokens(tokens, journey.getActivity().getName());
        }
        if (journey.getLocation() != null) {
            addTokens(tokens, journey.getLocation().getName());
        }
        return tokens.stream().filter(CommunityJourneyMatcher::isSignificant).collect(Collectors.toCollection(LinkedHashSet::new));
    }

    private static Set<String> insightTokens(CommunityInsightEntity insight) {
        Set<String> tokens = new LinkedHashSet<>();
        addTokens(tokens, insight.getJourneyContext());
        addTokens(tokens, insight.getTitle());
        return tokens.stream().filter(CommunityJourneyMatcher::isSignificant).collect(Collectors.toCollection(LinkedHashSet::new));
    }

    private static void addTokens(Set<String> tokens, String raw) {
        if (raw == null || raw.isBlank()) {
            return;
        }
        Arrays.stream(raw.toLowerCase(Locale.ROOT).split("[|\\s,/\\-—]+"))
                .map(String::trim)
                .filter(s -> !s.isEmpty())
                .forEach(tokens::add);
    }

    private static String token(String raw) {
        return raw == null ? "" : raw.trim().toLowerCase(Locale.ROOT);
    }

    private static boolean isSignificant(String token) {
        if (token.length() < 2) {
            return false;
        }
        return switch (token) {
            case "the", "and", "for", "your", "with", "from", "that", "this", "day", "pre", "prep" -> false;
            default -> true;
        };
    }
}
